clarity/src/vm/callables.rs

```diff
@@ -300,3 +339,216 @@ impl FunctionIdentifier {
         }
     }
 }
+
+// Implicitly cast principals to traits and traits to other traits as needed,
+// recursing into compound types. This function does not check for legality of
+// these casts, as that is done in the type-checker. Note: depth of recursion
+// should be capped by earlier checks on the types/values.
+fn implicit_cast(type_sig: &TypeSignature, value: &Value) -> Result<Value> {
+    Ok(match (type_sig, value) {
+        (
+            TypeSignature::OptionalType(inner_type),
+            Value::Optional(OptionalData {
+                data: Some(inner_value),
+            }),
+        ) => Value::Optional(OptionalData {
+            data: Some(Box::new(implicit_cast(inner_type, inner_value)?)),
+        }),
+        (
+            TypeSignature::ResponseType(inner_types),
+            Value::Response(ResponseData { committed, data }),
+        ) => Value::Response(ResponseData {
+            committed: *committed,
+            data: Box::new(implicit_cast(
+                if *committed {
+                    &inner_types.0
+                } else {
+                    &inner_types.1
+                },
+                data,
+            )?),
+        }),
+        (
+            TypeSignature::SequenceType(SequenceSubtype::ListType(list_type)),
+            Value::Sequence(SequenceData::List(ListData {
+                data,
+                type_signature,
+            })),
+        ) => {
+            let mut values = Vec::new();
+            for elem in data {
+                values.push(implicit_cast(list_type.get_list_item_type(), elem)?);
+            }
+            Value::Sequence(SequenceData::List(ListData {
+                data: values,
+                type_signature: type_signature.clone(),
+            }))
+        }
+        (
+            TypeSignature::TupleType(tuple_type),
+            Value::Tuple(TupleData {
+                type_signature: _,
+                data_map,
+            }),
+        ) => {
+            let mut cast_data_map = BTreeMap::new();
+            for (name, field_value) in data_map {
+                let to_type = match tuple_type.get_type_map().get(name) {
+                    Some(ty) => ty,
+                    None => {
```

Comment:
If the typechecker has already done a pass, then is this code reachable?

Reply:
That's right, this should not be reachable if the type checker has passed. Should I do something different here?

Reply:
Maybe just add a comment.

Reply:
Added in 49b103d64